modules/fanAdapter.js

```diff
@@ -0,0 +1,227 @@
+import * as utils from '../src/utils.js';
+import MD5 from 'crypto-js/md5.js';
+import { ajax } from '../src/ajax.js';
+import { BANNER, NATIVE } from '../src/mediaTypes.js';
+import { config } from '../src/config.js';
+import { registerBidder } from '../src/adapters/bidderFactory.js';
+const BIDDER_CODE = 'freedomadnetwork';
+const BASE_URL = 'https://srv.freedomadnetwork.com';
+
+/**
+ * Get user id from bid request. if no user id module used, return a new uuid.
+ *
+ * @param {BidRequest} bidRequest
+ * @returns {String} userId
+ */
+function getUserId(bidRequest) {
+  return generateUserId();
+}
+
+/**
+ * Get browser language
+ *
+ * @returns {String} language
+ */
+function getLanguage() {
+  const lang = (navigator.languages && navigator.languages[0]) ||
+    navigator.language || navigator.userLanguage;
+  return lang ? lang.split('-')[0] : 'en';
+}
+
+/**
+ * Get device info
+ *
+ * @returns {Object}
+ */
+function getDevice() {
+  const device = config.getConfig('device') || {};
+
+  device.w = device.w || window.screen.width;
+  device.h = device.h || window.screen.height;
+  device.ua = device.ua || navigator.userAgent;
+  device.language = device.language || getLanguage();
+  device.dnt = typeof device.dnt === 'number'
+    ? device.dnt : (utils.getDNT() ? 1 : 0);
+
+  return device;
+}
+
+/**
+ * Build OpenRTB request from bidRequest and bidderRequest
+ *
+ * @param {BidRequest} bidRequest
+ * @param {BidderRequest} bidderRequest
+ * @returns {Request}
+ */
+function buildBidRequest(bid, bidderRequest) {
+  const userId = getUserId(bid);
+
+  const payload = {
+    id: bid.bidId,
+    tmax: bidderRequest.timeout,
+    placements: [bid.params.placementId],
+    test: config.getConfig('debug') ? 1 : 0,
+    device: getDevice(),
+    at: 2,
+    user: {
+      coppa: config.getConfig('coppa') ? 1 : 0,
+    }
+  }
+
+  const gdprConsent = utils.deepAccess(bidderRequest, 'gdprConsent');
+  if (!!gdprConsent && gdprConsent.gdprApplies) {
+    payload.user.gdpr = 1;
+    payload.user.consent = gdprConsent.consentString;
+  }
+
+  const uspConsent = utils.deepAccess(bidderRequest, 'uspConsent');
+  if (uspConsent) {
+    payload.user.usp = uspConsent;
+  }
+
+  return {
+    method: 'POST',
+    url: BASE_URL + '/pb/req',
+    data: JSON.stringify(payload),
+    options: {
+      contentType: 'application/json',
+      withCredentials: false,
+      customHeaders: {
+        'Accept-Language': 'en;q=10',
+        'Authorization': 'Bearer ' + userId
```

Comment:
this is odd

Reply:
Unfortunately, I cannot change this. Server side requires this interface.

Reply:
why?

Reply:
I have no control over this part of the service.